monitors/icat_monitor.py

```diff
@@ -0,0 +1,107 @@
+"""
+Monitor ICAT for the latest run on an instrument. If end of run monitor is out of sync
+then restart it.
+"""
+
+import datetime
+import logging
+import re
+
+from monitors.settings import ICAT_MON_LOG_FILE
+from utils.clients.icat_client import ICATClient
+
+
+logging.basicConfig(filename=ICAT_MON_LOG_FILE,
+                    level=logging.INFO,
+                    format='%(asctime)s %(levelname)s [%(filename)s:%(lineno)d] %(message)s')
+
+
+def get_run_number(file_name, instrument_prefix):
+    """
+    Extract the run number from a RAW or Nexus file
+    """
+    # Check that the file name conforms to the expected pattern
+    match = re.match("%s[0-9]*(.nxs|.raw)" % instrument_prefix, file_name, re.IGNORECASE)
+    if not match:
+        logging.error("Returned data file does not match expected pattern: %s", file_name)
+        return None
+
+    # Extract the run number
+    file_name = file_name.replace(instrument_prefix, '')
+    run_number = ''.join([s for s in file_name if s.isdigit()])
+    return run_number
+
+
+def get_cycle_dates(icat_client):
+    """
+    What cycles could the last run have been in?
+    If the search space isn't constrained in some way then it takes far too long
+    to sort the list of data files. Narrowing down the dates is part of this.
+    """
+    date = datetime.datetime.today().strftime("%Y-%m-%d")
+    logging.info("Getting nearest cycles to current date (%s)", date)
+    last_cycle = icat_client.execute_query("SELECT c.startDate FROM FacilityCycle c"
```

Comment:
You could reduce some code duplication here if you do something like:
```
find_cycle_query = "SELECT c.endDate FROM FacilityCycle c  WHERE {0} < c.startDate ORDER BY c.endDate {1} LIMIT 0,1"
last_cycle = icat_client.execute_query(find_cycle_query.format(date, 'DESC')
next_cycle = icat_client.execute_query(find_cycle_query.format(date, 'ASC')
```

Reply:
Not for this PR, but I wonder if we find we have alot of common ICAT queries e.g. `find cycle` if we should store these together somewhere.

Reply:
Notice that the two queries are slightly different. One is greater than, one less than. Also, the start date is used in one and the end date in the other for good reason. We don't want there to be a hard change in the dates when we cross into a new cycle, for example.

Reply:
Ah okay these are more subtle than I saw on first inspection